File: leetcode/Tree/leetcode-110/110.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    Boolean balance = true;

    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        treeDepth(root);
        return balance;
    }

    public int treeDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int left = treeDepth(root.left);
        int right = treeDepth(root.right);
        if (Math.abs(left - right) > 1) {
            balance = false;
        }
        int depth = Math.max(left, right) + 1;
        return depth;
    }
}

File: leetcode/Tree/leetcode-110/110.go
package tree

// TreeNode Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}
	_, res := helper(root)
	return res
}

func helper(root *TreeNode) (int, bool) {
	if root == nil {
		return 0, true
	}

	left, lb := helper(root.Left)
	right, rb := helper(root.Right)
	balance := lb && rb
	if left-right > 1 || right-left > 1 {
		balance = false
	}

	return 1 + max(left, right), balance
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
